command.go

```diff
@@ -576,16 +577,17 @@ func (c *Command) Traverse(args []string) (*Command, []string, error) {
 	flags := []string{}
 	inFlag := false
 
+	c.mergePersistentFlags()
+
 	for i, arg := range args {
 		switch {
 		// A long flag with a space separated value
 		case strings.HasPrefix(arg, "--") && !strings.Contains(arg, "="):
-			// TODO: this isn't quite right, we should really check ahead for 'true' or 'false'
-			inFlag = !hasNoOptDefVal(arg[2:], c.Flags())
+			inFlag = !hasDefaultValue(arg[2:], c.Flags())
 			flags = append(flags, arg)
 			continue
 		// A short flag with a space separated value
-		case strings.HasPrefix(arg, "-") && !strings.Contains(arg, "=") && len(arg) == 2 && !shortHasNoOptDefVal(arg[1:], c.Flags()):
+		case isShortFlagWithNoDefaultValue(arg, c.Flags()):
```

Comment:
I may well be wrong, but my first glance makes me believe this may not be right. For short flags one could do (pseudo code)
```
BoolFlagP("bool1", "a")
BoolFlagP("bool2", "b")
StringFlagP("string", "s")
```
and on the command line do:
`-abs hello`

And be perfectly correct. This code will, I think, look only at `a` will completely ignore `b` and `s` and won't handle `hello` at all. The other code you referenced does seem to better handle short flags.

Reply:
Ah yes, that still needs to be solved, however that's not related to the original issue in #615 , right? The bug was that the flags were not in `c.Flags()` because they weren't merged yet.

Reply:
This function does seem to have a number of unrelated problems. I guess the reason it kinda works is because the call the ParseFlags() near the end does the merge. Merging (like in 615) at least seems to give this function a chance to be better.

Can you add a bunch of new _test for Traverse as you try to clean things up here?

Reply:
There are already a few tests for this function. I can add more as issues are fixed, but I think it would be good to merge this now, since it does fix one of the issues.

I opened #618 to track the other issue you found.